### PR TITLE
MC Docking Cation-Anion Poses

### PR DESCRIPTION
**Pull request by pafervi**
This pull request includes updates introduced for the Monte Carlo docking creating a MinDistanceCationAnionFitness class that generates poses with the positive charge of docked molecules close to the acid site of the catalyst. Hence creating more chemically realistic processes

### CODE REVIEW

File: VOID/dockers/success.py
Comment (ajhoffman1229) [inline VOID/dockers/success.py:60]: Just to make sure that I'm understanding correctly: this function is translating Cartesian coordinates to fractional coordinates using Pymatgen's `Structure` class?

File: VOID/dockers/success.py
Comment (ajhoffman1229) [inline VOID/dockers/success.py:102]: This code might not follow best practices: normally, you don't want to modify the inputs to a function. It's not as efficient, but it would be better to create a copy of the input, modify the copy, and return the copy on the last line.

File: VOID/fitness/threshold.py
Comment (ajhoffman1229) [inline VOID/fitness/threshold.py:110]: I don't think I saw a parser or function with this name. Is it located elsewhere? (maybe I am misunderstanding what the parser is and does)

File: VOID/fitness/threshold.py
Comment (ajhoffman1229) [inline VOID/fitness/threshold.py:7]: Is this threshold in &#8491;? Keeping cation-anion distances below 2 &#8491; seems very restrictive to me, especially since the other threshold is 1.5, but maybe I am misinterpreting these threshold values.

**Review by ajhoffman1229 — CHANGES_REQUESTED** [review]
I think most of this code looks very good. Everything is documented properly. I had a few clarifying questions and one critical change (in `success.py`) that probably need to made before merging, but otherwise this has my stamp of approval.

File: VOID/fitness/threshold.py
Comment (pafervi) [reply VOID/fitness/threshold.py:7]: You're absouletly right, I set this as a default for trials, then moved to parse htvs arguments and forgot about it. Changed it to 3.5 as a default value

File: VOID/dockers/success.py
Comment (pafervi) [reply VOID/dockers/success.py:60]: I knew this might bring some trouble. The reason for the use of this function is that the software works as a periodic crystal, but for every iteration the coordinates of the guest are displaced over the previous guest coordinates, so it was generating a correct pose but with an unvisualizable format (see example below). Though for running later calculations this shouldn't matter, I introduced this function that rescales the output once the optimal pose is achieved and allows visualization inside and outside htvs by different softwares.

Output without reescaling function:
Host coords:       1  1  1 
                            2  2  2
                            3   3   3
Guest coords:  235 432 274
                         123  456 789

File: VOID/dockers/success.py
Comment (pafervi) [reply VOID/dockers/success.py:102]: I'm not sure that I completely get what you are saying. You mean substituting the `self.rescale(cpx)` in dock by the whole process inside the function? or ammending the rescale function so it does not modify the input cpx? 
(I'm mostly sure you mean option B, but better double check)

```

    def dock(self, attempts):
        cpx = Complex(self.host.copy(), self.guest.copy())

        for trial in range(attempts):
            cpx = self.run(cpx, 1)

            if self.fitness(cpx) >= 0:
                print(f"{trial + 1} attempts to success")
                cpx = self.rescale(cpx)
                return [cpx]

        return []
```

File: VOID/dockers/success.py
Comment (ajhoffman1229) [reply VOID/dockers/success.py:60]: Yes, I think I've seen issues like what you describe. When you pull up the structure in VESTA or ASE (for example), the molecules are far outside the unit cell that is being shown. And this function will translate those positions to the fractional coordinates, right?

File: VOID/dockers/success.py
Comment (ajhoffman1229) [reply VOID/dockers/success.py:102]: I did mean the second thing you asked. Right now, the `rescale` function you wrote modifies the input `complex`. You have something like this:
```python
def rescale(self, complex):
    ...
    complex.host = Structure(...)
    ...
    return complex
```

Best coding practices suggest that you should make a copy of the input and modify that instead, like this:
```python
def rescale(self, complex):
    cpx = complex.copy() # this might not work for Complex objects, you should be able to use copy.deepcopy() as an alternative
    cpx.host = Structure(...)
    return cpx
```

File: VOID/dockers/success.py
Comment (pafervi) [reply VOID/dockers/success.py:60]: Yes, it is only translating the molecule from where it ends up outside the unit cell range to where it is placed in the unit cell range. Not changing the result of the docked position achieved at all and allowing visualization.

File: VOID/dockers/success.py
Comment (pafervi) [reply VOID/dockers/success.py:102]: Cool, I'll ammend this and give it a test

Comment (pafervi) [reply VOID/dockers/success.py:102]: At the end .copy.deepcopy() did not work and .copy() itself worked fine when testing.
Closing this PR and opening another with the changes made incorporated

**Pull request closed without merging**
